File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    let start = 0, maxLen = 1;

    const expandAroundCenter = (left, right) => {
        while (left >= 0 && right < s.length && s[left] === s[right]) {
            left--;
            right++;
        }
        return right - left - 1; // The length of the palindrome
    };

    for (let i = 0; i < s.length; i++) {
        let len1 = expandAroundCenter(i, i);      // Odd-length palindromes
        let len2 = expandAroundCenter(i, i + 1);  // Even-length palindromes
        let len = Math.max(len1, len2);
        
        if (len > maxLen) {
            maxLen = len;
            start = i - Math.floor((len - 1) / 2);
        }
    }

    return s.slice(start, start + maxLen);
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
//Expand around center
class Solution {
    // Helper function to expand around the center
    // This function will return the length of the palindrome expanded around the center
    private int expandAroundCenter(String s, int left, int right) {
        // Expand as long as the characters match and we are within bounds
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;   // Move left pointer outwards
            right++;  // Move right pointer outwards
        }
        // Return the length of the palindrome (right - left - 1)
        return right - left - 1;
    }

    public String longestPalindrome(String s) {
        // Base case: If the string is empty or has only one character, return the string itself
        if (s == null || s.length() < 2) {
            return s;
        }

        int start = 0, maxLen = 1; // Variables to track the start index and max length of the palindrome

        // Loop over the string to treat each character (and each pair of characters) as the center
        for (int i = 0; i < s.length(); i++) {
            // Expand around single character (odd-length palindrome)
            int len1 = expandAroundCenter(s, i, i);
            // Expand around two characters (even-length palindrome)
            int len2 = expandAroundCenter(s, i, i + 1);

            // Find the longer of the two palindromes
            int len = Math.max(len1, len2);

            // If the found palindrome is longer than the current max length
            if (len > maxLen) {
                // Update the start index and max length of the palindrome
                start = i - (len - 1) / 2;
                maxLen = len;
            }
        }

        // Return the longest palindrome substring
        return s.substring(start, start + maxLen);
    }
}
